cpp/ql/src/semmle/code/cpp/ir/internal/CppType.qll

```diff
@@ -0,0 +1,500 @@
+        (
+          if ptmType.getBaseType().getUnspecifiedType() instanceof RoutineType
+          then result = any(NullPointerType t).getSize() * 2
+          else result = any(NullPointerType t).getSize()
+        )
+      )
+    )
+  )
+}
+
+private int getTypeSize(Type type) {
+  if exists(getTypeSizeWorkaround(type))
+  then result = getTypeSizeWorkaround(type)
+  else result = type.getSize()
+}
+
+/**
+ * Holds if an `IRBooleanType` with the specified `byteSize` should exist.
+ */
+predicate hasBooleanType(int byteSize) {
+  byteSize = getTypeSize(any(BoolType type))
+}
+
+private predicate isSigned(IntegralOrEnumType type) {
+  type.(IntegralType).isSigned() or
+  exists(Enum enumType |
+    enumType = type.getUnspecifiedType() and
+    (
+      enumType.getExplicitUnderlyingType().getUnspecifiedType().(IntegralType).isSigned() or
+      not exists(enumType.getExplicitUnderlyingType())  // Assume signed.
+    )
+  )
+}
+
+private predicate isSignedIntegerType(IntegralOrEnumType type) {
+  isSigned(type) and not type instanceof BoolType
+}
+
+private predicate isUnsignedIntegerType(IntegralOrEnumType type) {
+  not isSigned(type) and not type instanceof BoolType
+}
+
+/**
+ * Holds if an `IRSignedIntegerType` with the specified `byteSize` should exist.
+ */
+predicate hasSignedIntegerType(int byteSize) {
+  byteSize = any(IntegralOrEnumType type | isSignedIntegerType(type)).getSize()
+}
+
+/**
+ * Holds if an `IRUnsignedIntegerType` with the specified `byteSize` should exist.
+ */
+predicate hasUnsignedIntegerType(int byteSize) {
+  byteSize = any(IntegralOrEnumType type | isUnsignedIntegerType(type)).getSize()
+}
+
+/**
+ * Holds if an `IRFloatingPointType` with the specified `byteSize` should exist.
+ */
+predicate hasFloatingPointType(int byteSize) {
+  byteSize = any(FloatingPointType type).getSize()
+}
+
+private predicate isPointerIshType(Type type) {
+  type instanceof PointerType or
+  type instanceof ReferenceType or
+  type instanceof NullPointerType
+}
+
+/**
+ * Holds if an `IRAddressType` with the specified `byteSize` should exist.
+ */
+predicate hasAddressType(int byteSize) {
+  // This covers all pointers, all references, and because it also looks at `NullPointerType`, it
+  // should always return a result that makes sense for arbitrary glvalues as well.
+  byteSize = any(Type type | isPointerIshType(type)).getSize()
+}
+
+/**
+ * Holds if an `IRFunctionAddressType` with the specified `byteSize` should exist.
+ */
+predicate hasFunctionAddressType(int byteSize) {
+  byteSize = any(NullPointerType type).getSize() or // Covers function lvalues
+  byteSize = getTypeSize(any(FunctionPointerIshType type))
+}
+
+private predicate isBlobType(Type type) {
+  (
+    exists(type.getSize()) and // Only include complete types
+    (
+      type instanceof ArrayType or
+      type instanceof Class or
+      type instanceof GNUVectorType
+    )
+  ) or
+  type instanceof PointerToMemberType // PTMs are missing size info
```

Comment:
I looked at reworking the logic so that `OpaqueType` was the fallback, but it didn't turn out very clean. Instead, I've added a sanity test that says that every type that has a size should also have a `CppType`, with a couple commented exceptions. This test did turn up the fact that we don't treat `_Float32`, `_Float64x`, etc. as `FloatingPointType`, so I fixed up `Type.qll` to handle those cases.

Reply:
And it looks like `BlockType` becomes an `IRFunctionAddressType`, so that's not an `OpaqueType`.